File: NZSLDict/Classes/DiagramViewController.m
//
//  DiagramViewController.m
//  NZSL Dict
//
//  Created by Greg Hewgill on 25/04/13.
//
//

#import "DiagramViewController.h"

#import "AppDelegate.h"
#import "Dictionary.h"
#import "DiagramView.h"
#import "NZSLDict-Swift.h"

@interface DiagramViewController ()

@end

@implementation DiagramViewController {
    DictEntry *currentEntry;
    UISearchBar *searchBar;
    DiagramView *diagramView;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
        self.tabBarItem = [[UITabBarItem alloc] initWithTitle:@"Diagram" image:[UIImage imageNamed:@"hands"] tag:0];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(showEntry:) name:EntrySelectedName object:nil];
    }
    return self;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)loadView
{
    UIView *view = [[UIView alloc] initWithFrame:[UIScreen mainScreen].bounds];
    view.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    
    float top_offset = IOS7() ? 20 : 0;
    
    searchBar = [[UISearchBar alloc] initWithFrame:CGRectMake(0, top_offset, view.bounds.size.width, 44)];
    searchBar.autoresizingMask = UIViewAutoresizingFlexibleWidth;
    searchBar.delegate = self;
    [view addSubview:searchBar];
    
    diagramView = [[DiagramView alloc] initWithFrame:CGRectMake(0, top_offset+44, view.bounds.size.width, view.bounds.size.height-(top_offset+44))];
    [view addSubview:diagramView];
    
    self.view = view;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view.
    if ([self respondsToSelector:@selector(edgesForExtendedLayout)]) {
        self.edgesForExtendedLayout = UIRectEdgeNone;
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewDidAppear:(BOOL)animated
{
    [self showCurrentEntry];
}

- (void)showEntry:(NSNotification *)notification
{
    currentEntry = notification.userInfo[@"entry"];
    if (diagramView == nil) {
        return;
    }
    [self showCurrentEntry];
}

- (void)showCurrentEntry
{
    searchBar.text = currentEntry.gloss;
    [diagramView showEntry:currentEntry];
}

- (UIBarPosition)positionForBar:(id<UIBarPositioning>)bar
{
    return UIBarPositionTopAttached;
}

- (BOOL)searchBarShouldBeginEditing:(UISearchBar *)searchBar
{
    [self.delegate returnToSearchView];
    return NO;
}

@end

File: NZSLDict/Classes/DiagramViewController.swift
import UIKit

class DiagramViewController: UIViewController, UISearchBarDelegate {
    var delegate: ViewControllerDelegate!
    var currentEntry: DictEntry!
    var diagramView: DiagramView!

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        self.tabBarItem = UITabBarItem(title: "Diagram", image: UIImage(named: "hands"), tag: 0)
        NotificationCenter.default.addObserver(self, selector: #selector(DiagramViewController.showEntry(_:)), name: .entrySelectedName, object: nil)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    override func loadView() {
        let view: UIView = UIView(frame: UIScreen.main.bounds)
        view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        view.backgroundColor = UIColor(named: "app-background")
        
        diagramView = DiagramView(frame: CGRect(x: 0, y: 0, width: view.bounds.size.width, height: view.bounds.size.height).insetBy(dx: 16.0, dy: 16.0))
        view.addSubview(diagramView)
        self.view = view
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        if self.responds(to: #selector(getter: UIViewController.edgesForExtendedLayout)) {
            self.edgesForExtendedLayout = UIRectEdge()
        }
        
        if #available(iOS 15.0, *) {
            self.tabBarController?.tabBar.scrollEdgeAppearance = self.tabBarController?.tabBar.standardAppearance
        }
    }
    
    func selectSearchMode(_ sender: UISegmentedControl) {
            self.tabBarController?.selectedIndex = 0
    }

    override func viewDidAppear(_ animated: Bool) {
        self.showCurrentEntry()
    }

    @objc func showEntry(_ notification: Notification) {
        currentEntry = (notification.userInfo!["entry"] as! DictEntry)
        if diagramView == nil {
            return
        }
        self.showCurrentEntry()
    }

    func showCurrentEntry() {
        diagramView.showEntry(currentEntry)
    }
}
